File: tests/Aspire.Cli.Tests/TestServices/TestGitWorktree.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

namespace Aspire.Cli.Tests.TestServices;

internal static class TestGitWorktree
{
    public static string WriteLinkedWorktreeMetadata(
        string worktreeRoot,
        string commonGitDirectory,
        string worktreeName = "feature",
        bool useRelativePaths = false)
    {
        var adminDirectory = Path.Combine(Path.GetFullPath(commonGitDirectory), "worktrees", worktreeName);
        Directory.CreateDirectory(adminDirectory);

        var gitFilePath = WriteGitDirFile(worktreeRoot, adminDirectory, useRelativePaths);
        var backPointer = useRelativePaths
            ? Path.GetRelativePath(adminDirectory, gitFilePath)
            : gitFilePath;
        File.WriteAllText(Path.Combine(adminDirectory, "gitdir"), backPointer + Environment.NewLine);

        return adminDirectory;
    }

    public static string WriteGitDirFile(string checkoutRoot, string gitDirectory, bool useRelativePath = false)
    {
        var fullCheckoutRoot = Path.GetFullPath(checkoutRoot);
        var fullGitDirectory = Path.GetFullPath(gitDirectory);
        Directory.CreateDirectory(fullCheckoutRoot);
        Directory.CreateDirectory(fullGitDirectory);

        var gitFilePath = Path.Combine(fullCheckoutRoot, ".git");
        var pointer = useRelativePath
            ? Path.GetRelativePath(fullCheckoutRoot, fullGitDirectory)
            : fullGitDirectory;
        File.WriteAllText(gitFilePath, $"gitdir: {pointer}{Environment.NewLine}");

        return gitFilePath;
    }
}

File: extension/src/test/testGitWorktree.ts
import * as fs from 'fs';
import * as path from 'path';

export function writeLinkedWorktreeMetadata(
    worktreeRoot: string,
    commonGitDirectory: string,
    worktreeName = 'feature',
    useRelativePaths = false): string {
    const fullWorktreeRoot = path.resolve(worktreeRoot);
    const adminDirectory = path.join(path.resolve(commonGitDirectory), 'worktrees', worktreeName);
    fs.mkdirSync(adminDirectory, { recursive: true });

    const gitFilePath = writeGitDirFile(fullWorktreeRoot, adminDirectory, useRelativePaths);
    const backPointer = useRelativePaths
        ? path.relative(adminDirectory, gitFilePath)
        : gitFilePath;
    fs.writeFileSync(path.join(adminDirectory, 'gitdir'), `${backPointer}\n`);

    return adminDirectory;
}

export function writeGitDirFile(
    checkoutRoot: string,
    gitDirectory: string,
    useRelativePath = false): string {
    const fullCheckoutRoot = path.resolve(checkoutRoot);
    const fullGitDirectory = path.resolve(gitDirectory);
    fs.mkdirSync(fullCheckoutRoot, { recursive: true });
    fs.mkdirSync(fullGitDirectory, { recursive: true });

    const gitFilePath = path.join(fullCheckoutRoot, '.git');
    const pointer = useRelativePath
        ? path.relative(fullCheckoutRoot, fullGitDirectory)
        : fullGitDirectory;
    fs.writeFileSync(gitFilePath, `gitdir: ${pointer}\n`);

    return gitFilePath;
}
